File: generate-otio.js
const fs = require('fs');

const IMAGE_FILES = [
    "clip01.jpg", 
    "clip02.jpg", 
    "clip03.jpg", 
    "clip04.jpg", 
    "clip05.jpg"
];
const DESIRED_CLIP_SECONDS = 10; // Desired duration of each clip in seconds
const startTimeCode = "01:00:00:00"; // User-defined start timecode for the timeline
const frame_rate = 24; // Ensure frame rate matches DaVinci Resolve settings
const outputPath = "./generated_otio/generated.otio";

// Array of slide indexes where you want specific transitions applied
const transitionPositions = [
    {index: 1, type: "D", duration: 24},  // Dissolve with duration
    {index: 2, type: "WR"},               // Wipe Right (Edge Wipe with angle 90)
    {index: 3, type: "WU"},               // Wipe Up (Edge Wipe with angle 0)
    {index: 4, type: "WL"}                // Wipe Left (Edge Wipe with angle -90)
];

const DEFAULT_DISSOLVE_DURATION = 24; // Default duration for dissolve if not provided

// Wipe direction angles based on type
const wipeAngles = {
    'WU': 0,    // Wipe Up
    'WR': 90,   // Wipe Right
    'WL': -90,  // Wipe Left
    'WD': 180   // Wipe Down
};

// Function to convert startTimeCode into a RationalTime for global_start_time
function timecodeToRationalTime(timecode, fps) {
    const [hours, minutes, seconds, frames] = timecode.split(":").map(Number);
    const totalSeconds = hours * 3600 + minutes * 60 + seconds + frames / fps;
    return { 
        "OTIO_SCHEMA": "RationalTime.1",
        "value": Math.floor(totalSeconds * fps),
        "rate": fps 
    };
}

// Function to create TimeRange objects
function createTimeRange(start, duration, fps) {
    return {
        "OTIO_SCHEMA": "TimeRange.1",
        "start_time": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": start,
            "rate": fps
        },
        "duration": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": duration,
            "rate": fps
        }
    };
}

// Function to create Clip objects for OTIO
function createClip(name, duration, fps) {
    return {
        "OTIO_SCHEMA": "Clip.1",
        "name": name,
        "available_range": createTimeRange(0, duration, fps),
        "type": "Clip"
    };
}

// Function to create Dissolve transitions
function createDissolveTransition(name, in_offset, out_offset) {
    return {
        "OTIO_SCHEMA": "Transition.1",
        "name": name,
        "transition_type": "Dissolve",
        "in_offset": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": in_offset,
            "rate": frame_rate
        },
        "out_offset": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": out_offset,
            "rate": frame_rate
        },
        "type": "Transition"
    };
}

// Function to create Edge Wipe transitions
function createEdgeWipeTransition(name, type, in_offset, out_offset) {
    const angle = wipeAngles[type];
    return {
        "OTIO_SCHEMA": "Transition.1",
        "name": name,
        "transition_type": "Custom_Transition",
        "in_offset": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": in_offset,
            "rate": frame_rate
        },
        "out_offset": {
            "OTIO_SCHEMA": "RationalTime.1",
            "value": out_offset,
            "rate": frame_rate
        },
        "metadata": {
            "Resolve_OTIO": {
                "Effects": {
                    "Effect Name": "Edge Wipe",
                    "Enabled": true,
                    "Name": "Edge Wipe",
                    "Parameters": [
                        {
                            "Default Parameter Value": 0,
                            "Key Frames": {},
                            "Parameter ID": "angle",
                            "Parameter Value": angle,
                            "Variant Type": "Int",
                            "maxValue": 360.0,
                            "minValue": -360.0
                        },
                        {
                            "Default Parameter Value": 0.0,
                            "Key Frames": {
                                "0": {
                                    "Value": 0.0,
                                    "Variant Type": "Double"
                                },
                                "24": { // Assuming transition duration is 24 frames
                                    "Value": 1.0,
                                    "Variant Type": "Double"
                                }
                            },
                            "Parameter ID": "transitionCustomCurvesKeyframes",
                            "Parameter Value": 0.0,
                            "Variant Type": "Double",
                            "maxValue": 1.0,
                            "minValue": 0.0
                        }
                    ],
                    "Type": 13
                },
                "Transition Type": "Edge Wipe"
            }
        },
        "type": "Transition"
    };
}

// Start building the OTIO timeline object
let timeline = {
    "OTIO_SCHEMA": "Timeline.1",
    "metadata": {
        "Resolve_OTIO": {
            "Resolve OTIO Meta Version": "1.0"
        }
    },
    "name": "Generated Timeline",
    "global_start_time": timecodeToRationalTime(startTimeCode, frame_rate),
    "tracks": [
        {
            "OTIO_SCHEMA": "Track.1",
            "name": "Video Track",
            "kind": "Video",
            "children": [],
            "type": "Track"
        }
    ],
    "type": "Timeline.1"
};

// Add clips and transitions to the track
IMAGE_FILES.forEach((fileName, index) => {
    // Create a clip and add it to the timeline
    const clip = createClip(fileName, DESIRED_CLIP_SECONDS, frame_rate);
    timeline.tracks[0].children.push(clip);

    // Check if there's a transition for the current clip
    const transition = transitionPositions.find(t => t.index === index);
    if (transition) {
        const transitionDuration = transition.duration || DEFAULT_DISSOLVE_DURATION;
        const in_offset = Math.floor(transitionDuration / 2);
        const out_offset = Math.floor(transitionDuration / 2);

        let transitionObj = null;
        if (transition.type === "D") {
            transitionObj = createDissolveTransition("Dissolve", in_offset, out_offset);
        } else if (['WU', 'WR', 'WL', 'WD'].includes(transition.type)) {
            transitionObj = createEdgeWipeTransition("Edge Wipe", transition.type, in_offset, out_offset);
        }

        if (transitionObj) {
            timeline.tracks[0].children.push(transitionObj);
        }
    }
});

// Save the OTIO JSON to a file
fs.writeFileSync(outputPath, JSON.stringify(timeline, null, 4));

console.log(`OTIO file generated at ${outputPath}`);

File: generate-otio.py
import opentimelineio as otio
import os

# Configuration for clips, transition settings, and frame rate
IMAGE_FILES = [
    "clip01.jpg", 
    "clip02.jpg", 
    "clip03.jpg", 
    "clip04.jpg", 
    "clip05.jpg"
]
DEFAULT_CLIP_DURATION = 10  # Default duration of each clip in seconds
FRAME_RATE = 24

# Start timecode (user-defined)
start_time_code = "01:00:00:00"

# Clip settings: slide indexes, transition types (D = Dissolve, C = Cut, Wipe90 = Wipe at 90 degrees)
# Optional: Provide custom transition duration using tDuration and clip duration using cDuration
clips_settings = [
    { "index": 0, "type": None, "cDuration": 10 },   # First clip with default duration, no transition
    { "index": 1, "type": "D", "tDuration": 24, "cDuration": 12 },  # Dissolve with custom duration for clip 2
    { "index": 2, "type": "D", "cDuration": 8 },                   # Dissolve with default transition duration for clip 3
    { "index": 3, "type": "Wipe90", "tDuration": 24, "cDuration": 15 },  # Wipe with custom duration for clip 4
    { "index": 4, "type": "C", "cDuration": 10 }                    # Regular cut with default clip duration
]

# Default transition settings
DEFAULT_DISSOLVE_DURATION = 24  # Default dissolve duration (frames)

# Create a timeline
timeline = otio.schema.Timeline(name="Improved Timeline")
video_track = otio.schema.Track(name="Video Track", kind=otio.schema.TrackKind.Video)
timeline.tracks.append(video_track)

# Create a clip with a specified or default duration
def create_clip(name, duration):
    return otio.schema.Clip(
        name=name,
        source_range=otio.opentime.TimeRange(
            start_time=otio.opentime.RationalTime(0, FRAME_RATE),
            duration=otio.opentime.RationalTime(duration * FRAME_RATE, FRAME_RATE)
        )
    )

# Create transitions based on type and duration
def create_transition(transition_type, in_offset=None, out_offset=None):
    if transition_type == "D":
        # Dissolve transition (default or custom duration)
        in_offset = otio.opentime.RationalTime(in_offset if in_offset else DEFAULT_DISSOLVE_DURATION, FRAME_RATE)
        out_offset = in_offset
        return otio.schema.Transition(
            transition_type=otio.schema.TransitionTypes.SMPTE_Dissolve,
            in_offset=in_offset,
            out_offset=out_offset
        )
    elif transition_type == "C":
        # Cut transition (no need to define an actual transition)
        return None
    elif transition_type == "Wipe90":
        # Create an edge wipe transition with metadata for DaVinci Resolve
        transition = otio.schema.Transition(
            transition_type="Custom_Transition",  # Custom type for wipe
            in_offset=otio.opentime.RationalTime(in_offset if in_offset else DEFAULT_DISSOLVE_DURATION, FRAME_RATE),
            out_offset=otio.opentime.RationalTime(out_offset if out_offset else DEFAULT_DISSOLVE_DURATION, FRAME_RATE)
        )

        # Adding custom metadata for DaVinci Resolve to recognize the Edge Wipe with angle and curves
        transition.metadata["Resolve_OTIO"] = {
            "Effects": {
                "Effect Name": "Edge Wipe",
                "Enabled": True,
                "Name": "Edge Wipe",
                "Parameters": [
                    {
                        "Default Parameter Value": 0,
                        "Key Frames": {},
                        "Parameter ID": "angle",
                        "Parameter Value": 90,  # Custom angle set to 90 degrees
                        "Variant Type": "Int",
                        "maxValue": 360.0,
                        "minValue": -360.0
                    },
                    {
                        "Default Parameter Value": 0.0,
                        "Key Frames": {
                            "0": {
                                "Value": 0.0,
                                "Variant Type": "Double"
                            },
                            "24": {
                                "Value": 1.0,
                                "Variant Type": "Double"
                            }
                        },
                        "Parameter ID": "transitionCustomCurvesKeyframes",
                        "Parameter Value": 0.0,
                        "Variant Type": "Double",
                        "maxValue": 1.0,
                        "minValue": 0.0
                    }
                ],
                "Type": 13  # Specific to edge wipe in DaVinci Resolve
            },
            "Transition Type": "Edge Wipe"
        }
        return transition
    else:
        # Fallback to a cut if an unknown transition type is given
        return None

# Initialize clips and transitions
for clip_data in clips_settings:
    clip_index = clip_data["index"]
    clip_name = IMAGE_FILES[clip_index]
    clip_duration = clip_data.get("cDuration", DEFAULT_CLIP_DURATION)
    
    clip = create_clip(clip_name, clip_duration)
    video_track.append(clip)

    # For all clips after the first, add a transition if specified
    if clip_index > 0:
        transition_type = clip_data.get("type", "C")
        transition_duration = clip_data.get("tDuration", DEFAULT_DISSOLVE_DURATION)
        transition = create_transition(transition_type, in_offset=transition_duration)
        
        if transition:
            video_track.append(transition)

# Ensure the directory exists
output_directory = "generated_otio"
os.makedirs(output_directory, exist_ok=True)

# Save the timeline to a .otio file in the specified directory
output_file_path = os.path.join(output_directory, "generated.otio")
otio.adapters.write_to_file(timeline, output_file_path)

print(f"Timeline created and saved as '{output_file_path}'")
